ui/src/app/base/selectors/resourcepool/resourcepool.ts

```diff
@@ -1,58 +1,69 @@
 import { createSelector } from "@reduxjs/toolkit";
 
-const resourcepool = {};
+import { ResourcePool } from "app/base/reducers/resourcepool/types";
+import RootState from "app/base/reducers/types";
 
 /**
  * Returns all resource pools.
  * @param {Object} state - The redux state.
  * @returns {Array} A list of all resource pools.
  */
-resourcepool.all = (state) => state.resourcepool.items;
+const all = (state: RootState): ResourcePool[] => state.resourcepool.items;
 
 /**
  * Whether resource pools are loading.
  * @param {Object} state - The redux state.
  * @returns {Boolean} Resource pool loading state.
  */
-resourcepool.loading = (state) => state.resourcepool.loading;
+const loading = (state: RootState): boolean => state.resourcepool.loading;
 
 /**
  * Whether resource pools have been loaded.
  * @param {Object} state - The redux state.
  * @returns {Boolean} Resource pools loaded state.
  */
-resourcepool.loaded = (state) => state.resourcepool.loaded;
+const loaded = (state: RootState): boolean => state.resourcepool.loaded;
 
 /**
  * Returns resource pool errors.
  * @param {Object} state - The redux state.
  * @returns {Object} Resource pool errors state.
  */
-resourcepool.errors = (state) => state.resourcepool.errors;
+const errors = (state: RootState): unknown => state.resourcepool.errors;
 
 /**
  * Get the saving state.
  * @param {Object} state - The redux state.
  * @returns {Boolean} Whether resource pools are being saved.
  */
-resourcepool.saving = (state) => state.resourcepool.saving;
+const saving = (state: RootState): boolean => state.resourcepool.saving;
 
 /**
  * Get the saved state.
  * @param {Object} state - The redux state.
  * @returns {Boolean} Whether resource pools have been saved.
  */
-resourcepool.saved = (state) => state.resourcepool.saved;
+const saved = (state: RootState): boolean => state.resourcepool.saved;
 
 /**
  * Returns resource pools that matches given id
  * @param {Object} state - The redux state.
  * @param {Number} id - id of resource pool to return.
  * @returns {Object} Resource pool that matches id.
  */
-resourcepool.getById = createSelector(
-  [resourcepool.all, (state, id) => id],
+const getById = createSelector(
+  [all, (_: RootState, id: number) => id],
   (pools, id) => pools.find((pool) => pool.id === id)
 );
 
+const resourcepool = {
```

Comment:
Assigning the selectors to the exported object at the end is fine, but we should be consistent and add a card perhaps if we'd rather do it this way. It is less verbose, but it also means you can potentially forget to export a selector.

Reply:
The reason I've changed it is because TS was raising errors about trying to assign parameters to an empty object. I guess I could instead create the object in one go, so we don't have to worry about adding it to the export e.g.
``` typescript
const pod = {
  all: (state: RootState): Pod[] => state.pod.items,
  loaded: (state: RootState): boolean => state.pod.loaded,
  ...
};

Reply:
> The reason I've changed it is because TS was raising errors about trying to assign parameters to an empty object. I guess I could instead create the object in one go, so we don't have to worry about adding it to the export e.g.
> 
> ```ts
> const pod = {
>   all: (state: RootState): Pod[] => state.pod.items,
>   loaded: (state: RootState): boolean => state.pod.loaded,
>   ...
> };
> 
> export default pod;
> 

That looks good, let's do that!

Reply:
ahh actually that won't work for any composed selectors because they relies on those methods actually existing first, nevermind

Reply:
Right okay, this is fine, but we should remember to update the other selectors with this shape when we convert them to ts.